bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java

```diff
@@ -0,0 +1,550 @@
+        synchronized (this) {
+            for (AdorneHubController.ZoneIDFutureRecord stateCommand : stateCommands) {
+                if (stateCommand.zoneId == zoneId) {
+                    @SuppressWarnings("unchecked")
+                    Map<String, Object> hubMsgPropertyList = (Map<String, Object>) hubMsg.get(HUB_TOKEN_PROPERTY_LIST);
+                    ThingTypeUID deviceType;
+                    String deviceTypeStr = (String) hubMsgPropertyList.get(HUB_TOKEN_DEVICE_TYPE);
+                    if (deviceTypeStr.equals(HUB_TOKEN_SWITCH)) {
+                        deviceType = THING_TYPE_SWITCH;
+                    } else if (deviceTypeStr.equals(HUB_TOKEN_DIMMER)) {
+                        deviceType = THING_TYPE_DIMMER;
+                    } else {
+                        logger.debug("Unsupported device type {}", deviceTypeStr);
+                        continue;
+                    }
+                    AdorneDeviceState state = new AdorneDeviceState(zoneId,
+                            (String) hubMsgPropertyList.get(HUB_TOKEN_NAME), deviceType,
+                            (Boolean) hubMsgPropertyList.get(HUB_TOKEN_POWER),
+                            ((Double) hubMsgPropertyList.get(HUB_TOKEN_POWER_LEVEL)).intValue());
+                    stateCommand.future.complete(state);
+                    removeStateCommands.add(stateCommand);
+                }
+            }
+            stateCommands.removeAll(removeStateCommands);
+        }
+    }
+
+    /**
+     * The hub informs us about a zone's change in properties.
+     */
+    private void processMsgZonePropertiesChanged(Map<String, Object> hubMsg) {
+        int zoneId = ((Double) hubMsg.get(HUB_TOKEN_ZID)).intValue();
+        logger.debug("Zone properties changed for zone ID {} ", zoneId);
+        @SuppressWarnings("unchecked")
+        Map<String, Object> hubMsgPropertyList = (Map<String, Object>) hubMsg.get(HUB_TOKEN_PROPERTY_LIST);
+        boolean onOff = (Boolean) hubMsgPropertyList.get(HUB_TOKEN_POWER);
+        int brightness = ((Double) hubMsgPropertyList.get(HUB_TOKEN_POWER_LEVEL)).intValue();
+        synchronized (this) {
+            AdorneHubChangeNotify changeListener = this.changeListener;
+            if (changeListener != null) {
+                changeListener.stateChangeNotify(zoneId, onOff, brightness);
+            }
+        }
+    }
+
+    /**
+     * The hub sent a list of zones in response to a command.
+     */
+    private void processMsgListZone(Map<String, Object> hubMsg) {
+        List<Integer> zones = new ArrayList<>();
+        @SuppressWarnings("unchecked")
+        List<Map<String, Double>> hubMsgZoneIdMapList = (List<Map<String, Double>>) hubMsg.get(HUB_TOKEN_ZONE_LIST);
+        for (Map<String, Double> zoneIdMap : hubMsgZoneIdMapList) {
+            zones.add(zoneIdMap.get(HUB_TOKEN_ZID).intValue());
+        }
+        synchronized (this) {
+            for (CompletableFuture<List<Integer>> zoneCommand : zoneCommands) {
```

Comment:
zoneCommands.forEach(zoneCommand -> zoneCommand.complete(zones));

Please also check other occurences of for-loops where you only iterate over the list to execute a command on each member.

Reply:
Fixed all of them